chain/epoch_manager/src/validator_selection.rs

```diff
@@ -0,0 +1,820 @@
+use crate::shard_assignment::assign_shards;
+use near_primitives::epoch_manager::epoch_info::EpochInfo;
+use near_primitives::epoch_manager::{EpochConfig, RngSeed};
+use near_primitives::errors::EpochError;
+use near_primitives::types::validator_stake::ValidatorStake;
+use near_primitives::types::{
+    AccountId, Balance, NumShards, ProtocolVersion, ValidatorId, ValidatorKickoutReason,
+};
+use num_rational::Ratio;
+use std::cmp::{self, Ordering};
+use std::collections::{hash_map, BTreeMap, BinaryHeap, HashMap, HashSet};
+
+pub fn proposals_to_epoch_info(
+    epoch_config: &EpochConfig,
+    rng_seed: RngSeed,
+    prev_epoch_info: &EpochInfo,
+    proposals: Vec<ValidatorStake>,
+    mut validator_kickout: HashMap<AccountId, ValidatorKickoutReason>,
+    validator_reward: HashMap<AccountId, Balance>,
+    minted_amount: Balance,
+    next_version: ProtocolVersion,
+) -> Result<EpochInfo, EpochError> {
+    debug_assert!(
+        proposals.iter().map(|stake| stake.account_id()).collect::<HashSet<_>>().len()
+            == proposals.len(),
+        "Proposals should not have duplicates"
+    );
+
+    let num_shards = epoch_config.num_block_producer_seats_per_shard.len() as NumShards;
+    let min_stake_ratio = {
+        let rational = epoch_config.validator_selection_config.minimum_stake_ratio;
+        Ratio::new(*rational.numer() as u128, *rational.denom() as u128)
+    };
+    let max_bp_selected = epoch_config.num_block_producer_seats as usize;
+    let max_cp_selected = max_bp_selected
+        + (epoch_config.validator_selection_config.num_chunk_only_producer_seats as usize);
+    let minimum_validators_per_shard =
+        epoch_config.validator_selection_config.minimum_validators_per_shard as usize;
+    let mut stake_change = BTreeMap::new();
+    let mut fishermen = vec![];
+    let proposals = proposals_with_rollover(
+        proposals,
+        prev_epoch_info,
+        &validator_reward,
+        &validator_kickout,
+        &mut stake_change,
+        &mut fishermen,
+    );
+    let mut block_producer_proposals =
+        order_proposals(proposals.values().filter(|p| !p.is_chunk_only()).cloned());
+    let (block_producers, bp_stake_threshold) =
+        select_block_producers(&mut block_producer_proposals, max_bp_selected, min_stake_ratio);
+    let mut chunk_producer_proposals = order_proposals(proposals.into_iter().map(|(_, p)| p));
+    let (chunk_producers, cp_stake_threshold) = select_chunk_producers(
+        &mut chunk_producer_proposals,
+        max_cp_selected,
+        min_stake_ratio,
+        num_shards,
+    );
+
+    // since block producer proposals could become chunk producers, their actual stake threshold
+    // is the smaller of the two thresholds
+    let bp_stake_threshold = cmp::min(bp_stake_threshold, cp_stake_threshold);
+
+    // proposals remaining chunk_producer_proposals were not selected for either role
```

Comment:
This comment is very confusing. I can't tell if it's intended to be such and I just can't parse it, or if it has something misspelled :)